linux/mcompile/keymap/mc_kmxfile.cpp

```diff
@@ -1,297 +1,512 @@
-
 #include "mc_kmxfile.h"
+#include <typeinfo>
+
+#define CERR_None                                          0x00000000
+#define CERR_CannotAllocateMemory                          0x00008004
+#define CERR_UnableToWriteFully                            0x00008007
+#define CERR_SomewhereIGotItWrong                          0x00008009
+
+KMX_BOOL KMX_VerifyKeyboard(LPKMX_BYTE filebase, KMX_DWORD sz);
+
+LPKMX_KEYBOARD KMX_FixupKeyboard(PKMX_BYTE bufp, PKMX_BYTE base, KMX_DWORD dwFileSize);
+
+KMX_BOOL KMX_SaveKeyboard(LPKMX_KEYBOARD kbd, PKMX_WCHAR filename) {
+
+  FILE *fp;
+  fp = Open_File(filename, u"wb");
 
-KMX_DWORD TEST2;
+  if(fp == NULL)
+  {
+    KMX_LogError(L"Failed to create output file (%d)", errno);
+    return FALSE;
+  }
+
+  KMX_DWORD err = KMX_WriteCompiledKeyboard(kbd, fp, FALSE);
+  fclose(fp);
 
-static KMX_BOOL LoadKeyboardFile(LPKMX_STR fileName, LPKEYBOARD *lpKeyboard);
+  if(err != CERR_None) {
+    KMX_LogError(L"Failed to write compiled keyboard with error %d", err);
 
-KMX_BOOL VerifyKeyboard(LPKMX_BYTE filebase, KMX_DWORD sz);
+    std::u16string u16_filname(filename);
+    std::string s = string_from_u16string(u16_filname);
+
+    remove(s.c_str());
+    return FALSE;
+  }
 
-LPKEYBOARD FixupKeyboard(PKMX_BYTE bufp, PKMX_BYTE base, KMX_DWORD dwFileSize);
+  return TRUE;
+}
 
-/*void Err(wchar_t *s) {
-	LogError(L"LoadKeyboard: %s, last error = %d\n", s, GetLastError());
-}*/
-/*BOOL LoadKeyboard(LPWSTR fileName, LPKEYBOARD *lpKeyboard) {
-	DWORD sz;
-	LPBYTE buf;
-	HANDLE hFile;
-	LPKEYBOARD kbp;
-  PBYTE filebase;
+KMX_DWORD KMX_WriteCompiledKeyboard(LPKMX_KEYBOARD fk, FILE* hOutfile, KMX_BOOL FSaveDebug) {
+
+	LPKMX_GROUP fgp;
+	LPKMX_STORE fsp;
+	LPKMX_KEY fkp;
+
+	PKMX_COMP_KEYBOARD ck;
+	PKMX_COMP_GROUP gp;
+	PKMX_COMP_STORE sp;
+	PKMX_COMP_KEY kp;
+	PKMX_BYTE buf;
+	KMX_DWORD size, offset;
+	DWORD i, j;
+
+	// Calculate how much memory to allocate
+	size = sizeof(KMX_COMP_KEYBOARD) +
+			fk->cxGroupArray * sizeof(KMX_COMP_GROUP) +
+			fk->cxStoreArray * sizeof(KMX_COMP_STORE) +
+      //wcslen(fk->szName)*2 + 2 +
+			//wcslen(fk->szCopyright)*2 + 2 +
+			//wcslen(fk->szLanguageName)*2 + 2 +
+			//wcslen(fk->szMessage)*2 + 2 +
+      fk->dwBitmapSize;
+
+	for(i = 0, fgp = fk->dpGroupArray; i < fk->cxGroupArray; i++, fgp++) {
+    if(fgp->dpName)
+	    size += u16len(fgp->dpName)*2 + 2;
```

Comment:
Easier to understand and better portable is:

```suggestion
	    size += (u16len(fgp->dpName) + 1) * sizeof(KMX_WCHAR);
```

same below for the other strings

Reply:
OK done: see PR #11879